File: backend/app/models/ticket.py
import enum
import uuid
from datetime import datetime

from sqlalchemy import BigInteger, DateTime, Enum, ForeignKey, Index, Text, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db import Base
from app.models._mixins import OrganizationScopedMixin, TimestampMixin, uuid7_pk


class TicketCategory(enum.StrEnum):
    """Ticket category — the casavi taxonomy, 32 values across 7 groups.

    Each value carries its group prefix so a flat dropdown still groups
    naturally; the SPA renders these grouped via the metadata dict in
    app/services/ticket_categories.py.
    """

    # Allgemeines
    ALLGEMEIN_FRAGE = "ALLGEMEIN_FRAGE"
    ALLGEMEIN_KLINGEL = "ALLGEMEIN_KLINGEL"
    ALLGEMEIN_DOKUMENTE = "ALLGEMEIN_DOKUMENTE"
    ALLGEMEIN_ONBOARDING = "ALLGEMEIN_ONBOARDING"
    ALLGEMEIN_LOB = "ALLGEMEIN_LOB"
    ALLGEMEIN_RUECKRUF = "ALLGEMEIN_RUECKRUF"
    ALLGEMEIN_SCHLUESSEL = "ALLGEMEIN_SCHLUESSEL"
    ALLGEMEIN_TELEFONNOTIZ = "ALLGEMEIN_TELEFONNOTIZ"
    # Buchhaltung und Zahlungsverkehr
    BUCHHALTUNG_BANK_SEPA = "BUCHHALTUNG_BANK_SEPA"
    BUCHHALTUNG_BETRIEBSKOSTEN = "BUCHHALTUNG_BETRIEBSKOSTEN"
    BUCHHALTUNG_JAHRESABRECHNUNG = "BUCHHALTUNG_JAHRESABRECHNUNG"
    BUCHHALTUNG_BELEGE = "BUCHHALTUNG_BELEGE"
    BUCHHALTUNG_ABBUCHUNGEN = "BUCHHALTUNG_ABBUCHUNGEN"
    # Immobilienvertrieb
    VERTRIEB_BEWERTUNG = "VERTRIEB_BEWERTUNG"
    VERTRIEB_BERATUNG = "VERTRIEB_BERATUNG"
    VERTRIEB_INTERESSE = "VERTRIEB_INTERESSE"
    # Mietverwaltung
    MIETER_WECHSEL = "MIETER_WECHSEL"
    # Schadensmeldung
    SCHADEN_ALLGEMEIN = "SCHADEN_ALLGEMEIN"
    SCHADEN_BAUMANGEL = "SCHADEN_BAUMANGEL"
    SCHADEN_ELEMENTAR = "SCHADEN_ELEMENTAR"
    SCHADEN_FEUER = "SCHADEN_FEUER"
    SCHADEN_SCHAEDLINGE = "SCHADEN_SCHAEDLINGE"
    SCHADEN_STROM = "SCHADEN_STROM"
    SCHADEN_ABWASSER = "SCHADEN_ABWASSER"
    SCHADEN_WASSER = "SCHADEN_WASSER"
    # WEG Verwaltung
    WEG_ANFRAGE = "WEG_ANFRAGE"
    WEG_BESCHLUSSANTRAG = "WEG_BESCHLUSSANTRAG"
    WEG_LEGIONELLEN = "WEG_LEGIONELLEN"
    # Sonstiges
    SONSTIGES_DATEN = "SONSTIGES_DATEN"
    SONSTIGES_BESCHLUSSUMSETZUNG = "SONSTIGES_BESCHLUSSUMSETZUNG"
    SONSTIGES_ETV = "SONSTIGES_ETV"
    SONSTIGES_RELAY = "SONSTIGES_RELAY"
    SONSTIGES_STOERUNG = "SONSTIGES_STOERUNG"
    SONSTIGES_OTHER = "SONSTIGES_OTHER"


class TicketStatus(enum.StrEnum):
    NEU = "NEU"
    OFFEN = "OFFEN"
    WARTET_AUF_KUNDE = "WARTET_AUF_KUNDE"
    GESCHLOSSEN = "GESCHLOSSEN"


class TicketShareScope(enum.StrEnum):
    """Who else (beyond creator + Verwalter) can see + comment on this ticket.

    PRIVATE       — creator + Verwalter only (default; explicit opt-in required
                    to widen access)
    PARTICIPANTS  — also: every user in ticket_participants
    PROPERTY      — also: every user with a contract on `tickets.property_id`
                    (requires property_id set). Only the explicitly-named
                    participants get email notifications; property-scope
                    viewers see the ticket if they visit the portal but
                    don't get fan-out emails.
    """

    PRIVATE = "PRIVATE"
    PARTICIPANTS = "PARTICIPANTS"
    PROPERTY = "PROPERTY"


class TicketMessageSource(enum.StrEnum):
    """Where a ticket message originated.

    PORTAL — typed into the React portal or Jinja admin UI
    EMAIL  — arrived via the SES → SNS → /webhooks/email/inbound path
    """

    PORTAL = "PORTAL"
    EMAIL = "EMAIL"


class Ticket(OrganizationScopedMixin, TimestampMixin, Base):
    __tablename__ = "tickets"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid7_pk)
    property_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("properties.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    created_by_user_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="RESTRICT"),
        nullable=True,
        index=True,
    )
    # When set, this ticket was created via inbound email by a sender that
    # does NOT have a WHV-Portal account. Notifications go to this address;
    # registered-user replies via the portal still work normally.
    external_sender_email: Mapped[str | None] = mapped_column(Text, nullable=True)
    assignee_user_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    category: Mapped[TicketCategory] = mapped_column(
        Enum(TicketCategory, name="ticket_category"),
        nullable=False,
    )
    status: Mapped[TicketStatus] = mapped_column(
        Enum(TicketStatus, name="ticket_status"),
        nullable=False,
        default=TicketStatus.NEU,
        server_default=TicketStatus.NEU.value,
    )
    share_scope: Mapped[TicketShareScope] = mapped_column(
        Enum(TicketShareScope, name="ticket_share_scope"),
        nullable=False,
        default=TicketShareScope.PRIVATE,
        server_default=TicketShareScope.PRIVATE.value,
    )
    subject: Mapped[str] = mapped_column(Text, nullable=False)

    # Updated on every message insert so the queue sorts by activity, not by
    # the original ticket created_at.
    last_message_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )
    closed_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )

    __table_args__ = (
        # Admin queue: filter by status, sort by recency.
        Index(
            "ix_tickets_org_status_last_msg",
            "organization_id",
            "status",
            "last_message_at",
        ),
        # "My tickets" lookup for the portal.
        Index("ix_tickets_creator_status", "created_by_user_id", "status"),
    )


class TicketMessage(Base):
    """A single message in a ticket thread.

    `is_internal_note` separates Verwalter-only notes from owner-visible
    replies. Filtered out before serialization for non-Verwalter callers.
    """

    __tablename__ = "ticket_messages"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid7_pk)
    ticket_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("tickets.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    # NULL when the message arrived via email from a non-registered sender.
    # In that case external_sender_email captures the address.
    author_user_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="RESTRICT"),
        nullable=True,
        index=True,
    )
    external_sender_email: Mapped[str | None] = mapped_column(Text, nullable=True)
    body: Mapped[str] = mapped_column(Text, nullable=False)
    is_internal_note: Mapped[bool] = mapped_column(
        nullable=False,
        default=False,
        server_default="false",
    )
    source: Mapped[TicketMessageSource] = mapped_column(
        Enum(TicketMessageSource, name="ticket_message_source"),
        nullable=False,
        default=TicketMessageSource.PORTAL,
        server_default=TicketMessageSource.PORTAL.value,
    )
    # RFC 5322 Message-ID of the inbound email (if source=EMAIL). Used for
    # idempotency on retries + threading on outbound replies (In-Reply-To /
    # References headers).
    email_message_id: Mapped[str | None] = mapped_column(Text, nullable=True, unique=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )

    __table_args__ = (
        # Thread render: load messages of a ticket in chronological order.
        Index("ix_ticket_messages_thread", "ticket_id", "created_at"),
    )


class TicketParticipant(Base):
    """Many-to-many: users explicitly added as participants on a ticket.

    Verwalter access is implicit (no row needed). Creator access is implicit
    (`tickets.created_by_user_id`). This table only tracks the *additional*
    named viewers — they get email fan-out and can comment.
    """

    __tablename__ = "ticket_participants"

    ticket_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("tickets.id", ondelete="CASCADE"),
        primary_key=True,
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        primary_key=True,
    )
    added_by_user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="RESTRICT"),
        nullable=False,
        index=True,
    )
    added_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )

    __table_args__ = (
        # Lookup for "tickets I'm participating in" (future use; today only
        # consulted as a membership-check inside ticket access rules).
        Index("ix_ticket_participants_user", "user_id"),
    )


class TicketMessageAttachment(Base):
    """File attached to a ticket message — uploaded from the SPA reply
    form or extracted from an inbound email's MIME tree.

    Always belongs to exactly one message (CASCADE on delete). Org scope
    rides along via message → ticket → organization_id; we don't store
    organization_id directly to keep the table thin, and ticket-level
    queries already join in to filter scope.

    Storage convention matches `documents.storage_url`: `"local-disk:<suffix>"`
    means the bytes live at `{ticket_attachment_dir}/{id}{suffix}`. Same
    Hetzner OS migration plan (REQUIREMENTS.md §1.4d iter 2) applies —
    swapping later only touches the storage helper.
    """

    __tablename__ = "ticket_message_attachments"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid7_pk)
    ticket_message_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("ticket_messages.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    # Original filename from the upload / MIME Content-Disposition.
    # Sanitised on read (Path(filename).name) but stored as-is so the
    # download endpoint can hand the user back what they uploaded.
    filename: Mapped[str] = mapped_column(Text, nullable=False)
    mime_type: Mapped[str | None] = mapped_column(Text, nullable=True)
    size_bytes: Mapped[int] = mapped_column(BigInteger, nullable=False)
    # See class docstring for the "local-disk:<suffix>" convention.
    storage_url: Mapped[str] = mapped_column(Text, nullable=False)
    # User who uploaded this attachment. NULL when the file arrived via
    # an inbound email — `external_sender_email` on the parent message
    # then identifies the sender.
    uploaded_by_user_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )

File: ios/WHV/Tickets/Ticket.swift
// Ticket models — mirror backend's TicketResponse /
// TicketDetailResponse / TicketMessageResponse plus the
// TicketCategory + TicketStatus + TicketShareScope enums.
//
// Categories are the casavi taxonomy (32 values across 7 groups);
// labels + the German grouping live here so the new-ticket sheet
// can render a grouped picker without a metadata round-trip.

import Foundation

enum TicketStatus: String, Codable, CaseIterable {
    case neu = "NEU"
    case offen = "OFFEN"
    case wartetAufKunde = "WARTET_AUF_KUNDE"
    case geschlossen = "GESCHLOSSEN"

    /// LocalizedStringResource so the build extractor sees the
    /// values and `Text(_ resource:)` re-renders when the user
    /// toggles English in Settings. Same pattern as
    /// AssemblyStatus.label.
    var label: LocalizedStringResource {
        switch self {
        case .neu: return "Neu"
        case .offen: return "Offen"
        case .wartetAufKunde: return "Wartet auf Kunde"
        case .geschlossen: return "Geschlossen"
        }
    }

    /// True when the ticket is still in motion. Drives grouping in
    /// the list (Aktuell vs. Geschlossen).
    var isActive: Bool {
        self != .geschlossen
    }
}

enum TicketShareScope: String, Codable {
    case privateScope = "PRIVATE"
    case participants = "PARTICIPANTS"
    case property = "PROPERTY"
}

/// 32 categories — German labels matching the backend's
/// `ticket_categories.py`. Grouped via `group` so the new-ticket
/// picker mirrors the admin SPA's grouped select.
enum TicketCategory: String, Codable, CaseIterable {
    case allgemeinFrage = "ALLGEMEIN_FRAGE"
    case allgemeinKlingel = "ALLGEMEIN_KLINGEL"
    case allgemeinDokumente = "ALLGEMEIN_DOKUMENTE"
    case allgemeinOnboarding = "ALLGEMEIN_ONBOARDING"
    case allgemeinLob = "ALLGEMEIN_LOB"
    case allgemeinRueckruf = "ALLGEMEIN_RUECKRUF"
    case allgemeinSchluessel = "ALLGEMEIN_SCHLUESSEL"
    case allgemeinTelefonnotiz = "ALLGEMEIN_TELEFONNOTIZ"
    case buchhaltungBankSepa = "BUCHHALTUNG_BANK_SEPA"
    case buchhaltungBetriebskosten = "BUCHHALTUNG_BETRIEBSKOSTEN"
    case buchhaltungJahresabrechnung = "BUCHHALTUNG_JAHRESABRECHNUNG"
    case buchhaltungBelege = "BUCHHALTUNG_BELEGE"
    case buchhaltungAbbuchungen = "BUCHHALTUNG_ABBUCHUNGEN"
    case vertriebBewertung = "VERTRIEB_BEWERTUNG"
    case vertriebBeratung = "VERTRIEB_BERATUNG"
    case vertriebInteresse = "VERTRIEB_INTERESSE"
    case mieterWechsel = "MIETER_WECHSEL"
    case schadenAllgemein = "SCHADEN_ALLGEMEIN"
    case schadenBaumangel = "SCHADEN_BAUMANGEL"
    case schadenElementar = "SCHADEN_ELEMENTAR"
    case schadenFeuer = "SCHADEN_FEUER"
    case schadenSchaedlinge = "SCHADEN_SCHAEDLINGE"
    case schadenStrom = "SCHADEN_STROM"
    case schadenAbwasser = "SCHADEN_ABWASSER"
    case schadenWasser = "SCHADEN_WASSER"
    case wegAnfrage = "WEG_ANFRAGE"
    case wegBeschlussantrag = "WEG_BESCHLUSSANTRAG"
    case wegLegionellen = "WEG_LEGIONELLEN"
    case sonstigesDaten = "SONSTIGES_DATEN"
    case sonstigesBeschlussumsetzung = "SONSTIGES_BESCHLUSSUMSETZUNG"
    case sonstigesEtv = "SONSTIGES_ETV"
    case sonstigesRelay = "SONSTIGES_RELAY"
    case sonstigesStoerung = "SONSTIGES_STOERUNG"
    case sonstigesOther = "SONSTIGES_OTHER"

    /// Unknown values from the server decode as `.sonstigesOther`
    /// instead of failing the whole detail fetch — forward-compat
    /// with future categories.
    init(from decoder: Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = TicketCategory(rawValue: raw) ?? .sonstigesOther
    }

    /// Group key — used as a String identity for grouped()
    /// (matches the backend GROUPS_ORDER) AND for the
    /// localized header. `groupLabel` returns the user-facing
    /// LocalizedStringResource the picker section header reads.
    var group: String {
        switch self {
        case .allgemeinFrage, .allgemeinKlingel, .allgemeinDokumente,
             .allgemeinOnboarding, .allgemeinLob, .allgemeinRueckruf,
             .allgemeinSchluessel, .allgemeinTelefonnotiz:
            return "Allgemeines"
        case .buchhaltungBankSepa, .buchhaltungBetriebskosten,
             .buchhaltungJahresabrechnung, .buchhaltungBelege,
             .buchhaltungAbbuchungen:
            return "Buchhaltung"
        case .vertriebBewertung, .vertriebBeratung, .vertriebInteresse:
            return "Immobilienvertrieb"
        case .mieterWechsel:
            return "Mietverwaltung"
        case .schadenAllgemein, .schadenBaumangel, .schadenElementar,
             .schadenFeuer, .schadenSchaedlinge, .schadenStrom,
             .schadenAbwasser, .schadenWasser:
            return "Schadensmeldung"
        case .wegAnfrage, .wegBeschlussantrag, .wegLegionellen:
            return "WEG-Verwaltung"
        case .sonstigesDaten, .sonstigesBeschlussumsetzung, .sonstigesEtv,
             .sonstigesRelay, .sonstigesStoerung, .sonstigesOther:
            return "Sonstiges"
        }
    }

    var groupLabel: LocalizedStringResource {
        switch group {
        case "Allgemeines": return "Allgemeines"
        case "Buchhaltung": return "Buchhaltung"
        case "Immobilienvertrieb": return "Immobilienvertrieb"
        case "Mietverwaltung": return "Mietverwaltung"
        case "Schadensmeldung": return "Schadensmeldung"
        case "WEG-Verwaltung": return "WEG-Verwaltung"
        default: return "Sonstiges"
        }
    }

    var label: LocalizedStringResource {
        switch self {
        case .allgemeinFrage: return "Allgemeine Frage"
        case .allgemeinKlingel: return "Klingelschild"
        case .allgemeinDokumente: return "Dokumentenanfrage"
        case .allgemeinOnboarding: return "Onboarding"
        case .allgemeinLob: return "Lob"
        case .allgemeinRueckruf: return "Rückrufbitte"
        case .allgemeinSchluessel: return "Schlüssel"
        case .allgemeinTelefonnotiz: return "Telefonnotiz"
        case .buchhaltungBankSepa: return "Bankverbindung / SEPA"
        case .buchhaltungBetriebskosten: return "Betriebskosten"
        case .buchhaltungJahresabrechnung: return "Jahresabrechnung"
        case .buchhaltungBelege: return "Belege"
        case .buchhaltungAbbuchungen: return "Abbuchungen"
        case .vertriebBewertung: return "Immobilienbewertung"
        case .vertriebBeratung: return "Verkaufsberatung"
        case .vertriebInteresse: return "Kaufinteresse"
        case .mieterWechsel: return "Mieterwechsel"
        case .schadenAllgemein: return "Allgemeiner Schaden"
        case .schadenBaumangel: return "Baumangel"
        case .schadenElementar: return "Elementarschaden"
        case .schadenFeuer: return "Feuerschaden"
        case .schadenSchaedlinge: return "Schädlinge"
        case .schadenStrom: return "Stromausfall"
        case .schadenAbwasser: return "Abwasser"
        case .schadenWasser: return "Wasserschaden"
        case .wegAnfrage: return "WEG-Anfrage"
        case .wegBeschlussantrag: return "Beschlussantrag"
        case .wegLegionellen: return "Legionellenprüfung"
        case .sonstigesDaten: return "Datenänderung"
        case .sonstigesBeschlussumsetzung: return "Beschlussumsetzung"
        case .sonstigesEtv: return "ETV"
        case .sonstigesRelay: return "Weiterleitung"
        case .sonstigesStoerung: return "Störung"
        case .sonstigesOther: return "Sonstiges"
        }
    }

    /// Plain-String form for places that can't accept a
    /// LocalizedStringResource (widget payload writers etc.).
    /// Looks up the current locale's translation at call time.
    var labelString: String {
        String(localized: label)
    }

    /// Group order matches the backend's GROUPS_ORDER so the iOS
    /// picker and the admin SPA always agree.
    static let groupsOrder: [String] = [
        "Allgemeines",
        "Buchhaltung",
        "Immobilienvertrieb",
        "Mietverwaltung",
        "Schadensmeldung",
        "WEG-Verwaltung",
        "Sonstiges",
    ]

    static func grouped() -> [(group: String, items: [TicketCategory])] {
        groupsOrder.map { group in
            (group, allCases.filter { $0.group == group })
        }
    }
}

struct TicketMessageAttachment: Codable, Identifiable, Hashable {
    let id: String
    let filename: String
    let mime_type: String?
    let size_bytes: Int
    let created_at: Date
}

struct TicketMessage: Codable, Identifiable, Hashable {
    let id: String
    let ticket_id: String
    let author_user_id: String?
    let author_email: String?
    let body: String
    // Server-side split of an e-mail reply: fresh text vs. the quoted thread
    // below it. Optional so older cached payloads still decode; fall back to
    // `body` when absent.
    let visible_body: String?
    let quoted_body: String?
    let is_internal_note: Bool
    let created_at: Date
    let attachments: [TicketMessageAttachment]

    /// Text to render by default — the reply without its quoted history.
    var displayBody: String { visible_body ?? body }
}

struct TicketParticipant: Codable, Identifiable, Hashable {
    let user_id: String
    let email: String?
    let added_by_user_id: String?
    let added_at: Date

    var id: String { user_id }
}

/// Full detail with the message thread + participants embedded.
struct TicketDetail: Codable, Identifiable, Hashable {
    let id: String
    let property_id: String?
    let created_by_user_id: String?
    let assignee_user_id: String?
    let category: TicketCategory
    let status: TicketStatus
    let share_scope: TicketShareScope
    let subject: String
    let last_message_at: Date
    let created_at: Date
    let closed_at: Date?
    let property_name: String?
    let property_address: String?
    let creator_email: String?
    let creator_contact_label: String?
    let creator_contact_id_impower: Int?
    let external_sender_email: String?
    let messages: [TicketMessage]
    let participants: [TicketParticipant]
}
